src/libstd/sys/windows/fs.rs

```diff
@@ -567,23 +646,85 @@ pub fn remove_dir_all(path: &Path) -> io::Result<()> {
         // rmdir only deletes dir symlinks and junctions, not file symlinks.
         rmdir(path)
     } else {
-        remove_dir_all_recursive(path)
+        // canonicalize to get a `//?/`-path, which can handle files like `CON`
+        // and `morse .. .`,  and when a directory structure is so deep it needs
+        // long path names
+        let path = try!(path.canonicalize());
+        let base_dir = match path.parent() {
+            Some(dir) => dir,
+            None => return Err(io::Error::new(io::ErrorKind::PermissionDenied,
+                                              "can't delete root directory"))
+        };
+        try!(remove_dir_all_recursive(path.as_ref(), base_dir.as_ref(), 0));
+        Ok(())
     }
 }
 
-fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
+fn remove_dir_all_recursive(path: &Path, base_dir: &Path, mut counter: u64)
+        -> io::Result<u64> {
+    // On Windows it is not enough to just recursively remove the contents of a
+    // directory and then the directory itself. Deleting does not happen
+    // instantaneously, but is scheduled. So just before or after flagging a
+    // file for deletion, we move it to some `base_dir` to avoid races.
```

Comment:
This implementation is getting pretty complicated, could you expand this comment to explain all the cases we're trying to catch here? It'll be nice to have a big intro to what's going on below.